Add documentation for Github OIDC federation with Keycloak

lkatalin opened the pull request:
This documents setting up Github OIDC integration federated with Keycloak. We have to federate through Keycloak because Github does not support PKCE for OAuth, which means as I understand it that it can only be used without federation for GHA (non-human-identity) workflows.

I was not able to figure out a way to add the Mapper for the IDP via the `realm-cr.yaml` file. For now I have instructions on how to do this through the Keycloak console. If anyone knows how to do this, please give an update. It may not be possible with the Keycloak version used by the RHSSO Operator, though I am not sure.

I also wanted to ask was whether it is in line with best security practices in OpenShift to have a Oauth app's client secret pasted into an unencrypted, run-of-the-mill `realm-cr.yaml` file. What is normally done in this situation? Without this, they have to do it manually.

sabre1041 commented on configure-oidc.md:
```diff
@@ -75,3 +75,128 @@ When issuing a `cosign sign` command, add the flags to pass in the client secret
+
+From the Keycloak console, navigate to `Identity Providers` in the lefthand menu bar, and select the provider you have just created: `github`. Select the tab `Mappers` and on the right-hand side, click the `create` button.
+
+Set the field values as shown below. The 'name' can be anything, as this will be overwritten.
```
Recommend providing the values here as well

sabre1041 commented on configure-oidc.md:
```diff
@@ -75,3 +75,128 @@ When issuing a `cosign sign` command, add the flags to pass in the client secret
+  <img src="/images/github_callback.png" />
+</p>
+
+### Add IdentityProviderMapper
```
Can we configure this via the `KeycloakRealm` CR to automate the process?

lkatalin replied:
That's what I was trying to figure out. As far as I could tell, there was no way to do it with the RHSSO Operator. Adding an `IdentityProviderMapper` or `IdentityProviderMappers` key was either ignored or made the KeycloakRealm fail to come up. If you know of a way to do this, I would love to add it.

sabre1041 commented on configure-oidc.md:
````diff
@@ -75,3 +75,128 @@ When issuing a `cosign sign` command, add the flags to pass in the client secret
+$ cosign sign -y --fulcio-url=$FULCIO_URL --rekor-url=$REKOR_URL --oidc-issuer=$OIDC_ISSUER_URL $IMAGE
+```
+
+On the login page the pops up in the browser, make sure to select the Github login option at the bottom.
````
```suggestion
On the login page the pops up in the browser, make sure to select the GitHub login option at the bottom.
```

sabre1041 commented on configure-oidc.md:
```diff
@@ -75,3 +75,128 @@ When issuing a `cosign sign` command, add the flags to pass in the client secret
+
+### Sign the artifact
+
+Sign the artifact using the same commands listed in [`sign-verify.md`](https://github.com/securesign/sigstore-ocp/blob/release-1.0.beta/sign-verify.md):
```
Recommend using relative pathing which will always align to the branch that the file is currently on

sabre1041 commented on images/mapper2.png:
Recommend showing a screenshot from the "Add Identity Provider Mapper" page

sabre1041 commented on configure-oidc.md:
```diff
@@ -75,3 +75,128 @@ When issuing a `cosign sign` command, add the flags to pass in the client secret
+
+### Add IdentityProviderMapper
+
+From the Keycloak console, navigate to `Identity Providers` in the lefthand menu bar, and select the provider you have just created: `github`. Select the tab `Mappers` and on the right-hand side, click the `create` button.
```
```suggestion
From the Keycloak console, navigate to `Identity Providers` in the lefthand menu bar, and select the provider you have just created: `github`. Select the tab `Mappers` and on the right-hand side and click the `create` button.
```

lkatalin commented on configure-oidc.md:
```diff
@@ -15,7 +15,7 @@ Once created, the credentials should be available in the Google Cloud Console un
 ### Update Fulcio deployment configuration
 
-Once RHTAS is installed, issue
+Once [RHTAS is installed](https://github.com/securesign/sigstore-ocp/blob/main/tas-easy-install.sh), issue
```
That's fair. What if we point them to the deployment guide instead? https://access.redhat.com/documentation/en-us/red_hat_trusted_artifact_signer/2024-q1/html/deployment_guide/installing-red-hat-trusted-artifact-signer-on-openshift_deploy

lkatalin replied:
I'm actually going to remove this change in this PR as it's GCP related, but the same applies for the Github installation step.

lkatalin commented on configure-oidc.md:
````diff
@@ -75,3 +75,128 @@ When issuing a `cosign sign` command, add the flags to pass in the client secret
+
+You will need to log in. The username will be `admin`. You can retrieve the password with:
+```shell
+$ oc get secret/credential-keycloak -n keycloak-system -o yaml | awk '/ADMIN_PASSWORD:/ {print $2}' | base64 -d; printf "\n"
````
I like it. Though I do prefer the newline, I can also see the argument against it for a cleaner command.

lkatalin commented on configure-oidc.md:
````diff
@@ -75,3 +75,128 @@ When issuing a `cosign sign` command, add the flags to pass in the client secret
+$ oc get routes -n keycloak-system | awk '/keycloak/ {print $2}' | head -1
+keycloak-keycloak-system.apps.rosa.umebj-fwkt3-8j8.y6tr.p3.openshiftapps.com
+```
+Copy this route, adding `https://` in front, into your browser.
````
Do you mean the output of the command or the note about https? I've reworked it so please take a look (will push the new version shortly). I do think it's important to make a note about https in case they haven't done this before, as otherwise it will not work.

lkatalin replied:
Oh I see, your command adds that already. Neat! :tada:

lkatalin commented on configure-oidc.md:
```diff
@@ -75,3 +75,128 @@ When issuing a `cosign sign` command, add the flags to pass in the client secret
+
+#### Option 1: Create a KeycloakRealm in OpenShift on the CLI
+
+Create a Custom Resource file from which to create the KeycloakRealm that includes the Github identity provider:
```
Well, they would need to delete it, as far as I'm aware (described below), or create it with a different name.

lkatalin replied:
I'm going through a flow where they make one with a new name right now, to test that, but I think it is actually easier if they delete the existing one. I get this error
```
status:
  loginURL: ""
  message: 'failed to create realm: (409) 409 Conflict'
  phase: failing
  ready: false
```
on the new Keycloakrealm when I haven't deleted the pre-existing `sigstore` one - I think because the IDP already exists in a different Keycloakrealm, but I am not sure.

lkatalin commented on configure-oidc.md:
````diff
@@ -75,3 +75,128 @@ When issuing a `cosign sign` command, add the flags to pass in the client secret
+Make sure that you have no existing KeycloakRealm with the `metadata.name` specified in your `realm-cr.yaml` file. IMPORTANT: make sure you are okay with deleting your current `sigstore` KeycloakRealm.
+```shell
+$ oc delete keycloakrealm sigstore -n keycloak-system
+keycloakrealm.keycloak.org "sigstore" deleted
````
Not sure, see above

lkatalin replied:
Oh I see, you mean the output

lkatalin commented on configure-oidc.md:
```diff
@@ -75,3 +75,128 @@ When issuing a `cosign sign` command, add the flags to pass in the client secret
+        providerId: github
+        trustEmail: true
+        config:
+            clientId: <my client id> # add your client ID
```
Thanks, the spacing was hard to get right. Let me know if there's still something wrong with it in the new version

lkatalin commented on configure-oidc.md:
````diff
@@ -75,3 +75,128 @@ When issuing a `cosign sign` command, add the flags to pass in the client secret
+```
+
+#### Option 2: 
+You may also add the Github Identity provider to an existing KeycloakRealm through the Keycloak console by selecting `Identity Providers` in the lefthand menu bar and selecting `Add provider -> Github` from the drop down menu. See above for instructions on how to access the Keycloak console.
````
I've added a few more steps here.

lkatalin commented on configure-oidc.md:
```diff
@@ -75,3 +75,128 @@ When issuing a `cosign sign` command, add the flags to pass in the client secret
+
+### Log in to Keycloak console
+
+You will need to log in to the Keycloak console to complete some steps of this process. Once [RHTAS is installed](https://github.com/securesign/sigstore-ocp/blob/main/tas-easy-install.sh), you can do this by finding the route and admin password.
```
Going to still update this with a link TBD

lkatalin commented:
Thanks for the review, @sabre1041 !

I think I've addressed your comments, except:
- Linking out to the RH Deployment guide for installation, rather than `./tas-easy-install.sh` - thoughts?
- Deleting the `sigstore` Keycloakrealm vs. creating a new one
- Whether we can automate adding the Mapper in the CR. This would be ideal

sabre1041 commented:
> Thanks for the review, @sabre1041 !
> 
> I think I've addressed your comments, except:
> 
> * Linking out to the RH Deployment guide for installation, rather than `./tas-easy-install.sh` - thoughts?

If we have the document included within the repository, it would be ideal as it avoids getting out of sync

> * Deleting the `sigstore` Keycloakrealm vs. creating a new one

While the opreator "should" make the appropriate changes with the updates specified, it might be easier to just delete and recreate.

> * Whether we can automate adding the Mapper in the CR. This would be ideal

Take a look at the `.spec.realm.identityProviderMappers`. there is usually a way to associate it within your`identityProvider` config

lkatalin commented:
> Take a look at the .spec.realm.identityProviderMappers. there is usually a way to associate it within your identityProvider config

Do you mean [this key](https://github.com/keycloak/keycloak-operator/blob/main/deploy/crds/keycloak.org_keycloakrealms_crd.yaml#L939)? I've tried to use this key in the `.yaml` but it resulted in errors. Sending you a message on this to see if you know what's going on there.

lkatalin commented:
I've made some updates:

- I added the Mapper to the custom resource file
- I removed the link to the TAS installation process, since there isn't an overview available in the beta release version of the repo
- I discovered the that Homepage URL field must also be updated in the OAuth app, so I added instructions for that

Let me know what you think, @sabre1041 .

Rendered version: https://github.com/lkatalin/sigstore-ocp/blob/github-oidc/configure-oidc.md#github-federated-via-keycloak

sabre1041 commented on configure-oidc.md:
```diff
@@ -75,3 +75,191 @@ When issuing a `cosign sign` command, add the flags to pass in the client secret
+
+Make sure to click `Update Application`.
+
+### Sign the artifact
```
Before signing the artifact, the `sigstore` client must be available within the `sigstore` realm. If you are creating the `sigstore` realm from scratch, it will not have the `sigstore` client unless you previously installed RHTAS which included both the realm and client and then created the `sigstore` realm via the CLI. it would then get recreated by the operator.

Regardless, there does need to be some wording that describes steps that the user needs to take regardless of the approach

lkatalin replied:
This is a good point, I'm putting a note in about the assumptions re: installation and will ping to see what you think.

lkatalin commented on configure-oidc.md:
````diff
@@ -75,3 +75,191 @@ When issuing a `cosign sign` command, add the flags to pass in the client secret
+Sign the artifact using the same commands listed in [`sign-verify.md`](/sign-verify.md):
+
+```shell
+$ cosign initialize --mirror=$TUF_URL --root=$TUF_URL/root.json
````
It resolves to this https://github.com/securesign/sigstore-ocp/blob/main/tas-easy-install.sh#L67 as long as they're running the `./tas-easy-install.sh` script. I agree we should either make that assumption explicit or give them other steps to take

sabre1041 replied:
The end user still needs to run `source ./tas-env-variables.sh` as described upon completion of the `./tas-easy-install.sh` script

lkatalin replied:
Setting up the env vars is described just above that by linking out to the `sign-verify.md` doc, but I've just added another line to say that they can also do it by using the post-install script.

lkatalin commented on configure-oidc.md:
```diff
@@ -1,6 +1,6 @@
 
-While RHTAS uses the RHSSO operator (based on Keycloak) by default, other OIDC options are available. Integrating with these options is described below. These steps should be followed after RHTAS installation.
+While RHTAS uses the RHSSO operator (based on Keycloak) for OIDC by default, other OIDC options are available. Integrating with these options is described below. These steps should be followed after [RHTAS installation](https://access.redhat.com/documentation/en-us/red_hat_trusted_artifact_signer/2024-q1/html/deployment_guide/index), as they will assume that other components, including the Keyclo ...
 
```
@sabre1041 I added this to point them to TAS installation steps.

lkatalin commented on configure-oidc.md:
```diff
@@ -75,3 +75,192 @@ When issuing a `cosign sign` command, add the flags to pass in the client secret
+
+Follow the commands listed in [`sign-verify.md`](/sign-verify.md) to ensure that your TAS environment variables are populated, then initialize Cosign and sign the artifact.
+**Troubleshooting:** If you have initialized Cosign previously, you may need to first delete `~/.sigstore` before running the commands below.
+
```
@sabre1041 I updated this as well to specify how to populate the env vars.

lkatalin commented:
Also - not sure if Konflux tests should be passing? I only changed markdown and added some images, so it shouldn't change anything relevant to that.

sabre1041 commented on configure-oidc.md:
```diff
@@ -1,6 +1,6 @@
 # OIDC configuration
 
-While RHTAS uses the RHSSO operator (based on Keycloak) by default, other OIDC options are available. Integrating with these options is described below. These steps should be followed after RHTAS installation.
+While RHTAS uses the RHSSO operator (based on Keycloak) for OIDC by default, other OIDC options are available. Integrating with these options is described below. These steps should be followed after [RHTAS installation](https://access.redhat.com/documentation/en-us/red_hat_trusted_artifact_signer/2024-q1/html/deployment_guide/index), as they will assume that other components, including the Keyclo ...
```
```suggestion
While RHTAS uses the RHSSO operator (based on Keycloak) for OIDC by default, other OIDC options are available. Integrating with these options is described below. These steps should be followed after [RHTAS installation](https://access.redhat.com/documentation/en-us/red_hat_trusted_artifact_signer/2024-q1/html/deployment_guide/index), as they will assume that other components, including the Keycloak client for RHTAS, is present, and that the environment variables for Cosign initialization are [populated](https://github.com/securesign/sigstore-ocp/blob/main/sign-verify.md#signing-a-container-from-the-local-system). You may need to recreate the Keycloak client for RHTAS once the `sigstore` realm has been fully configured as described within this guide.
```

sabre1041 reviewed (changes requested):
A few remaining changes. Otherwise, we should be good to go!

lkatalin commented:
@sabre1041 Done!

Though, separately, still not sure if the tests should be failing. Also... the openshift-ci bot seems to think I approved my own PR, which I didnt.

sabre1041 commented:
/lgtm
/approve

lkatalin commented:
The new version of the PR going into `release-1.0.gamma` is here https://github.com/securesign/sigstore-ocp/pull/188. Closing this in favor of that one.

lkatalin commented:
Reopening because this is the correct branch after all.

lkatalin commented:
Closing because of too many changes in the base repo.
